### PR TITLE
Adding AdvancedSetting to disable Color Primary Conversion of bt.2020 colors

### PR DESCRIPTION
**Pull request by Ablinne**
## Description
This adds a new entry to the AdvancedSettings video section that passes through rgb values w.r.t. bt.2020 primary colors on displays/connections that do not have the ability to transport the metadata that would normally be required. This disables the color primary conversion for bt.2020 primary colors. Currently, it only works for the LinuxRendererGL.

## Motivation and context
This is useful to watch wide gamut content in a context where the metadata can not be transported to the display and thus the renderer would otherwise default to a color primary conversion to bt709 primaries. When the color primaries of the source can be manually set on the connected display, this conversion can be bypassed.

## How has this been tested?
I have tested this change by visually comparing the results to modified video files where I have patched the color primary metadata. Say I have a video original.mp4 with bt2020 primaries. I patch the metadata using ffmpeg to bt709 primaries without changing the video content. This yields "wrong" colors, until i manually set the primaries of my display (Epson projector) to bt2020, restoring the correct colors. The same result can be achieved using the original.mp4 file in conjunction with this patch by disabling the color primary conversion. As long as the setting is not changed, the behaviour of the renderer does not change.

## What is the effect on users?
Users may notice an additional setting in their UI. Unless they change it, there is no effect.

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by Ablinne** [comment]
It seems for some reason one of builds is unable to checkout the source tree, so it fails. Not sure if I can do anything about it.
I have requested a wiki account to add some info about the new setting to it.
I am unsure what i can do about supporting this new setting in the other renderers, as i would probably be unable to test it.
Also i am not sure about what level of writing unit tests could be required.

**Comment by Ablinne** [comment]
I just noticed that due to the order of `#if` and `#endif` directives in `xbmc/system/shaders/GL/4.0/gl_yuv2rgb_filter4.glsl`, that tonemapping is not applied, when color conversion is disabled. I think technically this relation is not necessary, as color conversion and tonemapping really are separate operations. Enabling tonemapping without color conversion would be as easy as moving one `#endif` a few lines up in the shader definition(s), e. g. https://gist.github.com/Ablinne/9043d24de4b010f556685e5ec7865ce2. Anyone here have an opinion on that? Should I add that change, too?

**Comment by fuzzard** [comment]
I'll ping @sarbes for that, he might have some insight for you.

**Comment by thexai** [comment]
This seems to be in a context where, due to setup limitations, passing HDR metadata to the display and tonemap to SDR is not supported.

Being a limitation of the setup and not something that depends on each specific video, I don't understand why it is required to save a different configuration for each video in the VideoDatabase.

That is, my thought is that someone who needs to use this is because it has a very peculiar setup and will always need to use it (in all HDR videos) so it should be a player setting in GUI settings (not video/database settings).

I don't know the current state of Linux and GL renderer but this is certainly not needed on Windows. Even in setups that do not support HDR it is always possible to tonemap to SDR **and taking into account HDR metadata to do it**.

Note that passing HDR video untouched (BT2020 / transfer PQ) but without HDR associated metadata will not always result in the correct picture. Possibly the result will be very acceptable when the metadata is Max ML = 1000 nits only.

In short: I see it better as a setting in "System / Player" only visible in Linux platform.

**Comment by Ablinne** [comment]
> This seems to be in a context where, due to setup limitations, passing HDR metadata to the display and tonemap to SDR is not supported.

Yes and no. This is about not being able to pass the wide gamut information (primary colors), but not about the transfer function or tone mapping. This is a separate issue.

> Being a limitation of the setup and not something that depends on each specific video, I don't understand why it is required to save a different configuration for each video in the VideoDatabase.

There is still the case with SD Videos with bt601 primaries. I still want those converted to bt709. So i want this specifically for videos with bt2020 primaries.

> That is, my thought is that someone who needs to use this is because it has a very peculiar setup and will always need to use it (in all HDR videos) so it should be a player setting in GUI settings (not video/database settings).

If this global setting was limited to only disable color conversion if the source primaries are bt2020, I could live with that.

> I don't know the current state of Linux and GL renderer but this is certainly not needed on Windows. Even in setups that do not support HDR it is always possible to tonemap to SDR **and taking into account HDR metadata to do it**.

> Note that passing HDR video untouched (BT2020 / transfer PQ) but without HDR associated metadata will not always result in the correct picture. Possibly the result will be very acceptable when the metadata is Max ML = 1000 nits only.

My display (Epson EH-TW7100) has separate manual settings for the color primaries as well as the transfer function. So I can decide on a video by video basis if I want to have tonemapping in software or pass PQ scaled data to the display. This is a separate issue from the color primaries. I am aware that the maxCLL and maxFALL metadata are not passed in that situation and that those would be required for perfect playback, but in the practice they don't seem to make that much of a difference for _most_ content. I have even found reports that they are not even set correctly on a number of retail UHD disks.

> In short: I see it better as a setting in "System / Player" only visible in Linux platform.

With the mentioned restriction (not affecting bt601->bt709 conversion) I can live with that.

**Comment by Ablinne** [comment]
I have added the change allowing Tonemapping to be used independently of Color Primary Conversion. The LinuxRendererGL and LinuxRendererGLES already only activate tonemap, if the source has light metadata.

**Comment by Ablinne** [comment]
I have thought about the question of VideoSetting vs player setting a bit more. As explained I went for the VideoSetting, because I want to disable conversion it for bt2020 colors, but not bt601 colors and so I can decide on a per-video basis. In case of a player setting it would need to be clear that it affects bt2020 colors but not bt601 colors, so a concise naming would need to be found. Another option would be to somehow specify a full mapping of src primaries to dst primaries or a per video setting of the dst primaries, but this would be a relatively high effort.

If there is an agreement that the player setting is the better place and a suitable name can be found with the bt601 restriction and someone could give me some implementation hints (I fear it would not be as straight forward as a video setting where i could just replicate stuff from other settings), I could invest the time to move the setting. Otherwise I would prefer to leave it as it is.

The video setting could also be ported to the other renderers (to be respected in case no real hdr passthrough is available) or hidden from the ui for other renderes via a new entry in the ERENDERFEATURE enumeration. I might not be able to test this for the other renderers to a full extent so I'd rather leave this to someone who'd be able to, but let me know if I should go the route with ERENDERFEATURE to hide it in case of the other renderers.

**Comment by Ablinne** [comment]
@thexai  After thinking about it I came to agree that I would like to change this to a "setting in "System / Player" only visible in Linux platform" like you proposed, with the name "Passthrough BT.2020 RGB Primaries". It would only affect BT.2020 to BT.709 color conversion and nothing else (i.e. leave BG470M to BT.709 conversion in place). I will update this PR once I get it done. Can you hint me to what part of the code I would have to modify? It took me a while  to piece together on my own where all the parts of the VideoSettings need to go. A little hint might speed things up...

**Comment by Ablinne** [comment]
I have added an AdvancedSetting instead of a regular Setting. I believe this is an obscure enough feature such that it doesn't need to be in the GUI. Currently I have nothing to add to this and I hope this can be merged.

**Comment by lrusak** [comment]
Are you sure this even works properly? The drm subsystem in linux may only output bt702 unless the driver exposes the ability to send bt2020 in which case it likely has to be manually set. 

This can be done on the connector and the output plane, however only some hardware supports it.

```
        104 Colorspace:
                flags: enum
                enums: Default=0 SMPTE_170M_YCC=1 BT709_YCC=2 XVYCC_601=3 XVYCC_709=4 SYCC_601=5 opYCC_601=6 opRGB=7 BT2020_CYCC=8 BT2020_RGB=9 BT2020_YCC=10 DCI-P3_RGB_D65=11 DCI-P3_RGB_Theater=12
                value: 0

```

So be aware that you may not actually be outputting bt2020. If you can verify that your tv/screen is receiving bt2020 that would be the best. My tv has an indicator that shows this.

Something like this 
image

**Comment by Ablinne** [comment]
Yes I am using Linux. This setting is specifically meant to be a workaround for the case when it is _not_ possible to send properly flagged bt.2020 colors. In that case Kodi defaults to converting the RGB values to bt.709 primaries, because this is the default that most displays will assume.
This new setting will cause Kodi to skip this conversion and output RGB values w.r.t bt.2020 primaries anyway, relying on the user to tell the display manually what primaries to use. Not all displays allow the user to do so, but mine (Epson EH-TW7100) does.
I have tested this and it works just fine.

I am not sure why the most recent pipeline build failed, I have just updated the branch to include the style fixes that the pipeline had created on the previous (successful) run. All the changes were only whitespace.

**Comment by garbear** [comment]
Sometimes when you force push, jenkins gets confused and errs when pulling the new version. You can see win32 and freebsd had problems with the force push this time: https://jenkins.kodi.tv/job/BuildMulti-PR/23437/ (macos had a segfault running the test suite but it rebuilt automatically and this time it worked)

Team members have a magic incantation to rebuild: "jenkins build this please"

I think you can also trigger a rebuild by closing and reopening the PR.

**Comment by lrusak** [comment]
> Yes I am using Linux. This setting is specifically meant to be a workaround for the case when it is _not_ possible to send properly flagged bt.2020 colors. In that case Kodi defaults to converting the RGB values to bt.709 primaries, because this is the default that most displays will assume.
> 
> This new setting will cause Kodi to skip this conversion and output RGB values w.r.t bt.2020 primaries anyway, relying on the user to tell the display manually what primaries to use. Not all displays allow the user to do so, but mine (Epson EH-TW7100) does.
> 
> I have tested this and it works just fine.
> 

Fine as in the tv reports that it is receiving a bt2020 signal? Otherwise there isn't much value in this PR as it will likely just confuse people.

I've added support to set the DRM connector property in my branch here, however it will only work for kodi-gbm as Wayland/x11 don't have a way to do this yet (AFAIK). https://github.com/lrusak/xbmc/commit/aea35a53c44195dccbf36506802b64c1a0dad585



@Ablinne i ask because unless it is set explicitly in the DRM driver there will be an internal conversion happening anyways. So even if you think you are disabling the conversion in Kodi the DRM driver may be doing a conversion internally to what the output pipeline is configured for.

**Comment by Ablinne** [comment]
> > Yes I am using Linux. This setting is specifically meant to be a workaround for the case when it is _not_ possible to send properly flagged bt.2020 colors. In that case Kodi defaults to converting the RGB values to bt.709 primaries, because this is the default that most displays will assume.
> > This new setting will cause Kodi to skip this conversion and output RGB values w.r.t bt.2020 primaries anyway, relying on the user to tell the display manually what primaries to use. Not all displays allow the user to do so, but mine (Epson EH-TW7100) does.
> > I have tested this and it works just fine.
> 
> Fine as in the tv reports that it is receiving a bt2020 signal? Otherwise there isn't much value in this PR as it will likely just confuse people.
> 
> I've added support to set the DRM connector property in my branch here, however it will only work for kodi-gbm as Wayland/x11 don't have a way to do this yet (AFAIK). [lrusak@aea35a5](https://github.com/lrusak/xbmc/commit/aea35a53c44195dccbf36506802b64c1a0dad585)
> 
> @Ablinne i ask because unless it is set explicitly in the DRM driver there will be an internal conversion happening anyways. So even if you think you are disabling the conversion in Kodi the DRM driver may be doing a conversion internally to what the output pipeline is configured for.

No, the TV does not report it is getting a bt.2020 signal, it gets no metadata at all and assumes bt.709 by default. I have to tell it manually, that the RGB values it receives need to be interpreted as bt.2020. And by "it works fine" I mean that the combination of this new advancedsetting in Kodi and also telling my display "Hey, you're input is actually bt.2020, even though there is no metadata" results in colors that are correctly displayed.

I expect that the regular user would be confused by this, that's why it is an advanced setting that shouldn't be required for most users. Most users wouldn't be willing to manually change the setting on their display each time they switch between hd and 4k content.

The value in this PR is that one can watch wide gamut content even though the setup does not fully support it automatically, e.g. because X11 is used or the graphics card does not support the required HDMI version. Of course it requires manual intervention, but this may be acceptable to some people. So they can enable this advancedsetting and enjoy their content.

As far as I can see there is no further conversion by DRM or anything else, the RGB values that are created by the playback filters are passed to the display unchanged. I first experimented with a different approach where I patched the video files instead, which also worked and produced the same result. It's just more tedious.

**Comment by KarellenX** [comment]
@Ablinne 

Thanks for the edit to the wiki, but it is a bit premature to add the setting to the wiki page when the PR has yet to be accepted and merged. 
I hope you don't mind, but I am going to revert that edit for now. Once the PR is merged we can add it back, also with a note that it is available from v20 and which platforms it affects.
Let me know if you have any queries.
Thanks :)

**Comment by Ablinne** [comment]
> @Ablinne
> 
> Thanks for the edit to the wiki, but it is a bit premature to add the setting to the wiki page when the PR has yet to be accepted and merged. I hope you don't mind, but I am going to revert that edit for now. Once the PR is merged we can add it back, also with a note that it is available from v20 and which platforms it affects. Let me know if you have any queries. Thanks :)

Yes very well, thank you. I added it already because it was an item on the PR checklist which made it a bit unclear what the expected order of operations was. Adding the note about version and platform is a very good idea, will do that when the time comes.

**Comment by Ablinne** [comment]
Is there anything I can do and need to do for this to be able to get merged? I haven't checked in in a while, because I was confused about the expectations. Please let me know if more explanation is needed.
